128 extend controller to handle priorities

samvanderzwan opened the pull request:
The controller can now set the sources based on their priority. However the priority cannot yet be loaded form esdl since it does not know an attribute priority for producers. We will need to discuss with Edwin and Ewoud if this can be added.
Reviewing and merging of this branch is on hold until other PRs are handled. Then this branch will need some adjustments to be able to merge with master.

samvanderzwan commented:
This PR can also be reviewed, I also rebased it on current master branch

MichielTukker commented on ci/win32/profiling.cmd:
OK  to delete,  but what does this have to do with "Extend controller to handle priorities?"

Also,  it seems like a useful tool?

samvanderzwan replied:
nothing therefore I removed it after I added it by accident. And is certainly useful see #68

MichielTukker commented on doc/controller/controller.rst:
```diff
@@ -1,3 +1,31 @@
 =====================================
-To be filled.
+The controller is responsible for providing set points for the assets in the simulator.
+The controller parses the ESDL and creates objects of the assets it is providing set points for.
```
The controller parses ESDL?    The ESDL is parsed by ```simulator-core```  to create an object model of the ESDL which is used by our solver.  This parser now also creates the controller objects

samvanderzwan replied:
indeed, since the controller also needs information from the esdl

MichielTukker commented on doc/controller/controller.rst:
```diff
@@ -1,3 +1,31 @@
+for the controllable assets in the network. The key of this dict is the id of the asset.
+The value is another dict. The key of this dict is the property which needs to be set
+(e.g. supply temperature, heat demand). The value is the set point for this property.
+The controller works based on the priority of the source. The controller will first allocate
+The controller will also  if the sources with priority 1 have enough capacity to provide the demand.
```
Sentences are incomplete

samvanderzwan replied:
adjusted

MichielTukker commented on doc/controller/controller.rst:
```diff
@@ -1,3 +1,31 @@
+(e.g. supply temperature, heat demand). The value is the set point for this property.
+The controller works based on the priority of the source. The controller will first allocate
+The controller will also  if the sources with priority 1 have enough capacity to provide the demand.
+if more capacity is required, the sources with priority 2 will be used etc. If the demand is lower
+then the available capacity of the observed sources, the remaining demand will be equally distributed
+over the source at the observed priority.
```
So now we're implementing a very simple algorithm for the controller.   Does the current architecture allow for alternative control strategies?

samvanderzwan replied:
I added an abstract class which can serve as basis for other controllers

MichielTukker commented on src/simulator_core/entities/assets/controller/controller_classes.py:
File should be renamed to controller_abstract right?

samvanderzwan replied:
Yes, done

MichielTukker commented on src/simulator_core/entities/assets/controller/controller_consumer.py:
```diff
@@ -58,24 +49,24 @@ def get_heat_demand(self, time: datetime.datetime) -> float:
                 self.start_index = index
-                return float(self.profile["values"][index])
+                if self.profile["values"][index] > self.max_power:
+                    # TODO need to pass a message that power is insufficient and is
+                    #  capped to the max power available
```
Write the message to logging?

samvanderzwan replied:
doen

MichielTukker commented on src/simulator_core/entities/network_controller.py:
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
+        return producers
+
+    def _set_consumer_capped(self, time: datetime.datetime) -> dict:
+        """Method to set the consumer to the capped to the max avaialbl power of the producers."""
```
Spelling errors  +  undocumented function param

samvanderzwan replied:
fixed

MichielTukker commented on src/simulator_core/entities/network_controller.py:
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
+        factor = self.get_total_supply() / self.get_total_demand(time)
+        return self._set_consumer_to_demand(time=time, factor=factor)
+
+    def _set_consumer_to_demand(self, time: datetime.datetime, factor: float = 1.0) -> dict:
+        """Method to set the consumer to the demand."""
```
undocumented function parameters

samvanderzwan replied:
fixed

MichielTukker commented on src/simulator_core/entities/network_controller.py:
```diff
@@ -40,28 +41,84 @@ def run_time_step(self, time: datetime.datetime) -> dict:
+                                      PROPERTY_TEMPERATURE_SUPPLY: consumer.temperature_supply}
+        return consumers
+
+    def _set_producers_based_on_priority(self, time: datetime.datetime) -> dict:
+        """Method to set the producers based on priority."""
```
Undocumented function param

samvanderzwan replied:
fixed

MichielTukker commented on unit_test/entities/controller/test_controller_class.py:
```diff
@@ -0,0 +1,144 @@
+        self.assertEqual(producers[self.producer1.id][PROPERTY_HEAT_DEMAND], 1.0)
+        self.assertEqual(producers[self.producer2.id][PROPERTY_HEAT_DEMAND], 1.5)
+
+    def test_run_time_step(self):
+        """Test to run the time step."""
+        # Act
+        result = self.controller.run_time_step(datetime.now())
+
+        # Assert
+        self.assertIn(self.producer1.id, result)
+        self.assertIn(self.producer2.id, result)
+        self.assertIn(self.consumer1.id, result)
+        self.assertIn(self.consumer2.id, result)
```
Why not check the results?   this will be very valuable when the controller is improved further in the future  or the behavior/priorities are set in ESDL  (the results should not change)

samvanderzwan replied:
These are also tested in the other tests. It is usefull to add a test with actual parsing esdl and see what comes back

samvanderzwan replied:
added test with laoding of esdl and then creating controller and checking result for time step

samvanderzwan commented:
Made the adjustments, but also refactored the following;
1. Added abstract controller class, which in future can be used for other controllers
2. Changed the constructor of the controller class to be called with a factory method, which returns the producers and consumers as list. This also allows for easy extension of this class.

MichielTukker reviewed (changes requested):
tests don't run and still see linting errors

MichielTukker reviewed (changes requested):
tests don't run and still see linting errors

samvanderzwan commented on src/simulator_core/adapter/transforms/mappers.py:
```diff
@@ -204,23 +210,26 @@ def to_esdl(self, entity: NetworkController) -> EsdlObject:
         raise NotImplementedError("EsdlControllerMapper.to_esdl()")
 
-    def to_entity(self, model: EsdlObject) -> NetworkController:
+    def to_entity(self, model: int = 1) -> tuple[list[ControllerProducer], list[ControllerConsumer]]:
```
tests have been fixed, still need to discuss how to handle this best

samvanderzwan commented on src/simulator_core/entities/assets/controller/controller_consumer.py:
```diff
@@ -16,22 +16,16 @@
                                                            DEFAULT_TEMPERATURE_DIFFERENCE)
+from simulator_core.entities.assets.controller.controller_abstract_clas import (
+    AssetControllerAbstract)
+from simulator_core.entities.assets.esdl_asset_object import EsdlAssetObject
```
is similar as we do in heatnetwork pipe asset

samvanderzwan commented on src/simulator_core/entities/assets/esdl_asset_object.py:
```diff
@@ -59,6 +59,41 @@ def get_profile(self) -> pd.DataFrame:
                 return get_data_from_profile(esdl_port.profile[0])
         raise ValueError("No profile found for asset: " + self.esdl_asset.name)
 
+    def get_supply_temperature(self, port_type: str) -> float:
```
not entirely then you need to add other input argument to check on, so I decided to have two separate methods knowing we have duplicate code.
